File: hockey-react-frontend/src/services/PlayerService.js
/*
 * Centennial College - Winter 2026
 * Student: Engracia Batista
 * Student Number: 301394018
 *
 * PlayerService.js
 * Service layer for making HTTP calls to the HockeyPlayerService REST API.
 * Base URL targets the Spring Boot service running on port 8082.
 */

const PLAYER_API_URL = 'http://localhost:8082/player';

const PlayerService = {

    // GET all players
    getAllPlayers() {
        return fetch(PLAYER_API_URL)
            .then(response => response.json());
    },

    // GET a single player by ID
    getPlayerById(id) {
        return fetch(`${PLAYER_API_URL}/${id}`)
            .then(response => response.json());
    },

    // GET all players on a specific team
    getPlayersByTeam(teamId) {
        return fetch(`${PLAYER_API_URL}/team/${teamId}`)
            .then(response => response.json());
    },

    // GET all players at a position
    getPlayersByPosition(position) {
        return fetch(`${PLAYER_API_URL}/position/${position}`)
            .then(response => response.json());
    },

    // POST a new player
    addPlayer(player) {
        return fetch(PLAYER_API_URL, {
            method: 'POST',
            headers: {
                'Content-Type': 'application/json'
            },
            body: JSON.stringify(player)
        }).then(response => response.json());
    },

    // PUT to update an existing player
    updatePlayer(id, player) {
        return fetch(`${PLAYER_API_URL}/${id}`, {
            method: 'PUT',
            headers: {
                'Content-Type': 'application/json'
            },
            body: JSON.stringify(player)
        }).then(response => response.json());
    },

    // DELETE a player by ID
    deletePlayer(id) {
        return fetch(`${PLAYER_API_URL}/${id}`, {
            method: 'DELETE'
        }).then(response => response.json());
    }
};

export default PlayerService;

File: HockeyPlayerService/src/main/java/com/engracia/hockey/player/PlayerService.java
/*
 * Centennial College - Winter 2026
 * Student: Engracia Batista
 * Student Number: 301394018
 *
 * PlayerService.java
 * Business logic layer for the Hockey Player microservice.
 * All methods return reactive types (Mono or Flux) for non-blocking I/O.
 */
package com.engracia.hockey.player;

import java.util.Objects;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import lombok.AllArgsConstructor;
import reactor.core.publisher.Flux;
import reactor.core.publisher.Mono;

@Service
@Transactional
@AllArgsConstructor
public class PlayerService {

    @Autowired
    private PlayerRepository playerRepository;

    // Return all players; emit empty flux if none exist
    public Flux<Player> getAll() {
        return playerRepository.findAll().switchIfEmpty(Flux.empty());
    }

    // Return a single player by their ID
    public Mono<Player> getById(final int id) {
        return playerRepository.findById(id);
    }

    // Return all players on a specific team
    public Flux<Player> getByTeamId(final int teamId) {
        return playerRepository.findByTeamId(teamId).switchIfEmpty(Flux.empty());
    }

    // Return all players at a given position
    public Flux<Player> getByPosition(final String position) {
        return playerRepository.findByPosition(position).switchIfEmpty(Flux.empty());
    }

    // Save a new player to MongoDB
    public Mono<Player> save(final Player player) {
        return playerRepository.save(player);
    }

    // Update an existing player record
    public Mono<Player> update(final int id, final Player player) {
        return playerRepository.save(player);
    }

    // Delete a player by ID; returns the deleted player or empty if not found
    public Mono<Player> delete(final int id) {
        final Mono<Player> dbPlayer = getById(id);
        if (Objects.isNull(dbPlayer)) {
            return Mono.empty();
        }
        return getById(id)
            .switchIfEmpty(Mono.empty())
            .filter(Objects::nonNull)
            .flatMap(playerToDelete ->
                playerRepository.delete(playerToDelete).then(Mono.just(playerToDelete))
            );
    }
}
